Document best practices for position arguments in `utils.report()` (#8257)
stylelint/stylelint: docs/developer-guide/rules.md

Always use the:

- `validateOptions()` utility to warn users about invalid options
- `isStandardSyntax*()` utilities before checking a node or string to ignore non-standard syntax
- `report()` utility to report lint problems

##### Location arguments for `report()`

When using `report()`, you can specify the location of a problem in various ways:

- only `node` - implicitly span the entire range of the given node
- `word` - the first instance of a word in the serialized given node
- `index` and `endIndex` offsets - an index range within the given node

> - `start` and `end` positions - a position range within the given node

Author: I wasn't sure how to describe these. Any suggestions?

Reviewer: If you are talking about `start`/`end`, I would link to the PostCSS API directly.

i.e.
```suggestion
- `index` and `endIndex` offsets - an index range within the given node
- `start` and `end` positions - a [range](https://postcss.org/api/#range) within the given node
```

Each approach has pros and cons concerning:

- convenience - ease of use
- narrowness - how well a location matches the issue
- correctness - chance of bugs or incorrect locations
- performance - the cost of calcuating the location

For example, it's often convenient to use only `node` or use `word` but that comes at the expense of narrownesst and correctness.

In the following example of using `word`, the location is the selector within the rule:

```js
root.walkRules((ruleNode) => {
  const { selector } = ruleNode;

  report({
    result,
    ruleName,
    message: messages.rejected(),
    node: ruleNode,
    word: selector
  });
});
```

Using `word` is the slowest approach to finding a location within a node, especially on nodes with many children, e.g. at-rules.

> When using offsets and positions, you can use the `nodeFieldIndices` utilities, e.g. `declarationValueIndex`, to get the index of part of a node. These utilities account for the `raw` fields in the PostCSS AST.

Reviewer: If we are recommending our internal utilities we should probably add them to the `Utils` type and to the `stylelint.utils` chapter of `plugins.md`.

We didn't bikeshed about too the name of the file because we thought it would remain private.

Author: Let's move that discussion to another issue. This doc is primarily for rule authors.

> When using offsets and positions, you can use the `nodeFieldIndices` utilities, e.g. `declarationValueIndex`, to get the index of part of a node. These utilities account for the `raw` fields in the PostCSS AST.
> 
> In the following example of using `index`, `endIndex` and the `declarationValueIndex` utility, the location spans the value of the declaration:

Reviewer 2: [suggestion] This is minor, but how about making the consistent form for the function like `report()`?

```suggestion
When using offsets and positions, you can use the `nodeFieldIndices` utilities, e.g. `declarationValueIndex()`, to get the index of part of a node. These utilities account for the `raw` fields in the PostCSS AST.

In the following example of using `index`, `endIndex` and the `declarationValueIndex()` utility, the location spans the value of the declaration:
```

```js
root.walkDecls((declNode) => {
  const { prop, value } = declNode;

  const index = declarationValueIndex(decl);
  const endIndex = index + value.length;

  report({
    result,
    ruleName,
    message: messages.rejected(),
    node: declNode,
    index,
    endIndex:
  });
});
```

This approach also works well when using construct-specific parsers:

```js
root.walkDecls((declNode) => {

>   const { prop, value as declValue } = declNode;

Reviewer 2: [suggestion] We should fix the invalid syntax for `as`:

```suggestion
  const { prop, value: declValue } = declNode;
```

Author: Oops, we totally should! Good catch.

Reviewer: Should we use `remark-lint-code-block-syntax`?
related: #7983

Author: Let's open an issue to discuss.
